File: model-service/app/scoring/rules_engine.py
"""
model-service/app/scoring/rules_engine.py
规则引擎（T1，架构 §3.2 / §7.8 / 类图 RulesEngine）。

职责（仅本批次，不越界到 T4 完整评分卡装配）：
- load_rules(preset_id)：从 presets/*.json 加载规则（缺省回退 default.json）。
- flatten_dim_weight(stage, job_type, rules)：把「两层级权重」预折叠为
  扁平 dimWeight（Σ=1，仅含本阶段启用客观维）。
- compute_stage_score(objective, subjective, rules, stage[, job_type])：
  按 PRD §10 公式 + 架构 §3.2 计算客观分/主观分/总分/verdict。
- verdict_from_total(total[, rules, stage])：Q4 阈值映射。

零新增依赖（纯 Python + pydantic 的 Verdict 枚举复用）。
"""
from __future__ import annotations

import json
import os
from typing import Dict, List, Optional

from ..schemas import Verdict

_PRESETS_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "presets")


# ======================================================================
# 1) 规则加载
# ======================================================================
def load_rules(preset_id: str = "default") -> dict:
    """
    加载指定预设规则 JSON；若预设文件不存在则回退 default.json。
    找不到任何预设时抛 FileNotFoundError（上层捕获，不静默）。
    """
    target = os.path.join(_PRESETS_DIR, f"{preset_id}.json")
    if not os.path.isfile(target):
        target = os.path.join(_PRESETS_DIR, "default.json")
    if not os.path.isfile(target):
        raise FileNotFoundError(
            f"规则预设缺失：既无 {preset_id}.json 也无 default.json（期望目录 {_PRESETS_DIR}）"
        )
    with open(target, "r", encoding="utf-8") as fh:
        return json.load(fh)


# ======================================================================
# 2) 权重预折叠（架构 §7.8）
# ======================================================================
def _detect_job_type(objective: Dict[str, float]) -> str:
    """由 objective 中的 craft 维前缀推断工种（无 craft 维默认 code）。"""
    for dim in objective:
        if dim.startswith("img_"):
            return "image"
        if dim.startswith("txt_"):
            return "text"
        if dim.startswith("code_"):
            return "code"
    return "code"


def flatten_dim_weight(stage: str, job_type: str, rules: dict) -> Dict[str, float]:
    """
    按架构 §3.2 + §7.8 预折叠为扁平 dimWeight（Σ=1，仅含本阶段启用客观维）。

    公式：
      generic 块：genericRadarWeight[dim] × objectiveBlockWeight.generic
      craft   块：objectiveBlockWeight.craft 均分给 jobs[job_type].craftDims
      kpiRoi  块：performance 阶段预留（本批次无 kpi/roi 维 → 占位 0 维）
      最后整体归一化使 Σ=1（保证缺 kpiRoi 维时不会 <1）。

    约定（架构 §7.8）：__generic__ = genericRadar；__craft__ = 当前 jobType.craftDims。
    """
    stage_cfg = rules["stages"][stage]
    bw = stage_cfg.get("objectiveBlockWeight", {})
    generic_w = stage_cfg.get("genericRadarWeight", {})
    generic_block = float(bw.get("generic", 0.0))
    craft_block = float(bw.get("craft", 0.0))

    job_cfg = rules.get("jobs", {}).get(job_type, {})
    craft_dims = job_cfg.get("craftDims", [])

    raw: Dict[str, float] = {}
    # —— generic 块 ——
    for dim, w in generic_w.items():
        raw[dim] = float(w) * generic_block
    # —— craft 块（均分）——
    if craft_dims:
        per = craft_block / len(craft_dims)
        for d in craft_dims:
            raw[d] = per
    # —— kpiRoi 块：本批次占位（无维度），归一时自动把其份额
    #    重新分配给 generic + craft，保持 Σ=1（见下方归一）。

    total = sum(raw.values())
    if total <= 0:
        return raw
    return {k: v / total for k, v in raw.items()}


# ======================================================================
# 3) 阶段计分（PRD §10 公式 + 架构 §3.2）
# ======================================================================
def compute_stage_score(
    objective: Dict[str, float],
    subjective: Dict[str, float],
    rules: dict,
    stage: str,
    job_type: Optional[str] = None,
) -> dict:
    """
    计算单阶段评分卡核心（客观分/主观分/总分/verdict）。

    objectiveScore = Σ(objective[dim]/5 × dimWeight[dim]) × 100
      其中 dimWeight = flatten_dim_weight(stage, job_type, rules)
      （objective 中缺失的 dimWeight 维按 0 计；dimWeight 中多出的维忽略）
    subjectiveScore = Σ(subjective[dim]/5 × 1/|subjectiveDims|) × 100
      （主观维等权；enabledSubjective 来自 rules[stage]）
    total = objectiveScore × objectiveWeight(stage) + subjectiveScore × subjectiveWeight(stage)
    verdict = verdict_from_total(total, rules, stage)  （Q4：≥78 MVP / 50–78 OBSERVE / <50 FIRED）

    job_type 可显式传入；缺省时由 objective 中 craft 维前缀自动推断。
    """
    job_type = job_type or _detect_job_type(objective)
    stage_cfg = rules["stages"][stage]
    dim_weight = flatten_dim_weight(stage, job_type, rules)

    # 客观分（加权）
    obj_acc = 0.0
    for dim, w in dim_weight.items():
        score = float(objective.get(dim, 0.0))
        obj_acc += (score / 5.0) * w
    objective_score = round(obj_acc * 100.0, 1)

    # 主观分（等权）
    sub_dims = stage_cfg.get("enabledSubjective", [])
    n_sub = len(sub_dims) or 1
    sub_acc = 0.0
    for d in sub_dims:
        score = float(subjective.get(d, 0.0))
        sub_acc += (score / 5.0) * (1.0 / n_sub)
    subjective_score = round(sub_acc * 100.0, 1)

    # 总分
    ow = float(stage_cfg.get("objectiveWeight", 0.5))
    sw = float(stage_cfg.get("subjectiveWeight", 0.5))
    total = round(objective_score * ow + subjective_score * sw, 1)

    verdict = verdict_from_total(total, rules, stage)

    return {
        "objectiveScore": objective_score,
        "subjectiveScore": subjective_score,
        "total": total,
        "verdict": verdict,
        "jobType": job_type,
        "stage": stage,
        "dimWeight": dim_weight,
    }


# ======================================================================
# 4) verdict 映射（Q4）
# ======================================================================
def verdict_from_total(
    total: float,
    rules: Optional[dict] = None,
    stage: Optional[str] = None,
) -> Verdict:
    """
    total → Verdict（架构 §7.4 / Q4）：
      ≥ mvp(默认78)        → MVP
      78 > total ≥ observe(50) → OBSERVE
      < observe               → FIRED
    阈值优先取 rules[stage].thresholds（m vp / obse rve），否则用 Q4 默认。
    """
    mvp, observe = 78, 50
    if rules and stage and stage in rules.get("stages", {}):
        th = rules["stages"][stage].get("thresholds", {})
        mvp = float(th.get("mvp", mvp))
        observe = float(th.get("observe", observe))

    if total >= mvp:
        return Verdict.MVP
    if total >= observe:
        return Verdict.OBSERVE
    return Verdict.FIRED

File: src/engine/scoring/rulesEngine.ts
/**
 * src/engine/scoring/rulesEngine.ts
 * 规则引擎（T1 前端 TS 镜像，与 model-service/app/scoring/rules_engine.py 同公式）。
 *
 * 前端可离线计算（无需后端），公式与后端严格一致。
 * 零新增依赖（纯 TS）。
 */
import type { JobType, StageKey, SubjectiveDim, CraftDim } from "./registry";
import { JOB_GENERIC_WEIGHT } from "./registry";

/** 规则 JSON 的类型（与 presets/default.json 同构*/
export interface ScoringRules {
  $schema?: string;
  version?: string;
  presetId?: string;
  genericRadar?: string[];
  jobs: Record<JobType, { craftDims: CraftDim[] }>;
  stages: Record<
    StageKey,
    {
      enabledObjective?: string[];
      enabledSubjective?: SubjectiveDim[];
      objectiveBlockWeight: { generic: number; craft: number; kpiRoi?: number };
      objectiveWeight: number;
      subjectiveWeight: number;
      genericRadarWeight: Record<string, number>;
      thresholds?: { mvp: number; observe: number };
    }
  >;
  subjective?: { capPercent?: number; neutralBaseline?: number };
}

/** 由 objective 中的 craft 维前缀推断工种（无 craft 维默认 code） */
export function detectJobType(objective: Record<string, number>): JobType {
  for (const dim of Object.keys(objective)) {
    if (dim.startsWith("img_")) return "image";
    if (dim.startsWith("txt_")) return "text";
    if (dim.startsWith("code_")) return "code";
  }
  return "code";
}

/** 权重预折叠（镜像后端 flatten_dim_weight*/
export function flattenDimWeight(
  stage: StageKey,
  jobType: JobType,
  rules: ScoringRules,
): Record<string, number> {
  const stageCfg = rules.stages[stage];
  const bw = stageCfg.objectiveBlockWeight;
  const genericBlock = bw.generic ?? 0;
  const craftBlock = bw.craft ?? 0;

  const craftDims = rules.jobs[jobType]?.craftDims ?? [];

  // generic 六维权重：优先按工种差异化（Q2，JOB_GENERIC_WEIGHT[jobType]），
  // 缺失时回退阶段级 genericRadarWeight
  const genericW =
    (JOB_GENERIC_WEIGHT as Record<string, Record<string, number>>)[jobType] ??
    stageCfg.genericRadarWeight;

  const raw: Record<string, number> = {};
  // generic 块
  for (const [dim, w] of Object.entries(genericW)) {
    raw[dim] = w * genericBlock;
  }
  // craft 块（均分）
  if (craftDims.length > 0) {
    const per = craftBlock / craftDims.length;
    for (const d of craftDims) raw[d] = per;
  }
  // kpiRoi 块：本批次占位（无维度），归一时自动重分配其份额

  const total = Object.values(raw).reduce((a, b) => a + b, 0);
  if (total <= 0) return raw;
  const out: Record<string, number> = {};
  for (const [k, v] of Object.entries(raw)) out[k] = v / total;
  return out;
}

/** verdict 映射（Q4，镜像后端 verdict_from_total） */
export function verdictFromTotal(
  total: number,
  rules?: ScoringRules,
  stage?: StageKey,
): "MVP" | "OBSERVE" | "FIRED" {
  let mvp = 78;
  let observe = 50;
  if (rules && stage && rules.stages[stage]?.thresholds) {
    mvp = rules.stages[stage].thresholds!.mvp ?? mvp;
    observe = rules.stages[stage].thresholds!.observe ?? observe;
  }
  if (total >= mvp) return "MVP";
  if (total >= observe) return "OBSERVE";
  return "FIRED";
}

/** 阶段计分（镜像后端 compute_stage_score*/
export function computeStageScore(
  objective: Record<string, number>,
  subjective: Record<string, number>,
  rules: ScoringRules,
  stage: StageKey,
  jobType?: JobType,
): {
  objectiveScore: number;
  subjectiveScore: number;
  total: number;
  verdict: "MVP" | "OBSERVE" | "FIRED";
  jobType: JobType;
  stage: StageKey;
  dimWeight: Record<string, number>;
} {
  const jt = jobType ?? detectJobType(objective);
  const stageCfg = rules.stages[stage];
  const dimWeight = flattenDimWeight(stage, jt, rules);

  // 客观分（加权）
  let objAcc = 0;
  for (const [dim, w] of Object.entries(dimWeight)) {
    const score = objective[dim] ?? 0;
    objAcc += (score / 5) * w;
  }
  const objectiveScore = Math.round(objAcc * 100 * 10) / 10;

  // 主观分（等权）
  const subDims = stageCfg.enabledSubjective ?? [];
  const nSub = subDims.length || 1;
  let subAcc = 0;
  for (const d of subDims) {
    const score = subjective[d] ?? 0;
    subAcc += (score / 5) * (1 / nSub);
  }
  const subjectiveScore = Math.round(subAcc * 100 * 10) / 10;

  // 总分
  const ow = stageCfg.objectiveWeight ?? 0.5;
  const sw = stageCfg.subjectiveWeight ?? 0.5;
  const total = Math.round((objectiveScore * ow + subjectiveScore * sw) * 10) / 10;

  const verdict = verdictFromTotal(total, rules, stage);

  return {
    objectiveScore,
    subjectiveScore,
    total,
    verdict,
    jobType: jt,
    stage,
    dimWeight,
  };
}
